enginetest/enginetests.go

```diff
@@ -1700,6 +1701,507 @@ func TestTracing(t *testing.T, harness Harness) {
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t2 (pk) VALUES (1), (2), (3)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t2",
+		[]sql.Row{{1, 2}, {2, 2}, {3, 3}},
+	)
+
+	// Default expression converting to proper column type
+	TestQuery(t, harness, e,
+		"CREATE TABLE t3(pk BIGINT PRIMARY KEY, v1 VARCHAR(20) DEFAULT (GREATEST(pk, 2)))",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t3 (pk) VALUES (1), (2), (3)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t3",
+		[]sql.Row{{1, "2"}, {2, "2"}, {3, "3"}},
+	)
+
+	// Default literal of different type but implicitly converts
+	TestQuery(t, harness, e,
+		"CREATE TABLE t4(pk BIGINT PRIMARY KEY, v1 BIGINT DEFAULT '4')",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t4 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t4",
+		[]sql.Row{{1, 4}, {2, 4}},
+	)
+
+	// Back reference to default literal
+	TestQuery(t, harness, e,
+		"CREATE TABLE t5(pk BIGINT PRIMARY KEY, v1 BIGINT DEFAULT (v2), v2 BIGINT DEFAULT 7)",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t5 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t5",
+		[]sql.Row{{1, 7, 7}, {2, 7, 7}},
+	)
+
+	// Forward reference to default literal
+	TestQuery(t, harness, e,
+		"CREATE TABLE t6(pk BIGINT PRIMARY KEY, v1 BIGINT DEFAULT 9, v2 BIGINT DEFAULT (v1))",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t6 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t6",
+		[]sql.Row{{1, 9, 9}, {2, 9, 9}},
+	)
+
+	// Forward reference to default expression
+	TestQuery(t, harness, e,
+		"CREATE TABLE t7(pk BIGINT PRIMARY KEY, v1 BIGINT DEFAULT (8), v2 BIGINT DEFAULT (v1))",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t7 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t7",
+		[]sql.Row{{1, 8, 8}, {2, 8, 8}},
+	)
+
+	// Back reference to value
+	TestQuery(t, harness, e,
+		"CREATE TABLE t8(pk BIGINT PRIMARY KEY, v1 BIGINT DEFAULT (v2 + 1), v2 BIGINT)",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t8 (pk, v2) VALUES (1, 4), (2, 6)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t8",
+		[]sql.Row{{1, 5, 4}, {2, 7, 6}},
+	)
+
+	// TEXT expression
+	TestQuery(t, harness, e,
+		"CREATE TABLE t9(pk BIGINT PRIMARY KEY, v1 LONGTEXT DEFAULT (77))",
+		[]sql.Row(nil),
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t9 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t9",
+		[]sql.Row{{1, "77"}, {2, "77"}},
+	)
+
+	// DATETIME/TIMESTAMP NOW/CURRENT_TIMESTAMP literal
+	TestQuery(t, harness, e,
```

Comment:
You should add a TODO to come back to these tests that you can't check values on yet. We need to be able to set the clock for tests.

Reply:
Great idea!